File: apple/Sources/ReactNativeWatchOS/Modifiers/GridCellAnchor.swift
import SwiftUI

/// SwiftUI `.gridCellAnchor(_:)`. Positions a view within its `Grid` cell.
/// Accepts a named anchor or an explicit `{ x, y }` unit point. Available
/// on watchOS 9+.
enum RNWGridCellAnchorModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("gridCellAnchor") { view, params, _ in
            let anchor = unitPoint(params["anchor"]) ?? .center
            return AnyView(view.gridCellAnchor(anchor))
        }
    }

    /// Map a named anchor string or an `{ x, y }` dict to a `UnitPoint`.
    private static func unitPoint(_ raw: Any?) -> UnitPoint? {
        if let s = raw as? String {
            switch s {
            case "center":         return .center
            case "top":            return .top
            case "bottom":         return .bottom
            case "leading":        return .leading
            case "trailing":       return .trailing
            case "topLeading":     return .topLeading
            case "topTrailing":    return .topTrailing
            case "bottomLeading":  return .bottomLeading
            case "bottomTrailing": return .bottomTrailing
            default:               return nil
            }
        }
        if let dict = raw as? [String: Any] {
            let x = (dict["x"] as? NSNumber)?.doubleValue ?? 0
            let y = (dict["y"] as? NSNumber)?.doubleValue ?? 0
            return UnitPoint(x: x, y: y)
        }
        return nil
    }
}

File: src/modifiers/gridCellAnchor.ts
import { createModifier } from './createModifier';

/// A named anchor point, or an explicit unit point in the 0...1 space
/// (`{ x: 0, y: 0 }` is the top-leading corner, `{ x: 1, y: 1 }` is
/// bottom-trailing).
export type GridCellAnchor =
  | 'center'
  | 'top'
  | 'bottom'
  | 'leading'
  | 'trailing'
  | 'topLeading'
  | 'topTrailing'
  | 'bottomLeading'
  | 'bottomTrailing'
  | { x: number; y: number };

export interface GridCellAnchorParams {
  /// The unit point within the cell that the view is anchored to.
  anchor: GridCellAnchor;
}

/// SwiftUI `.gridCellAnchor(_:)`. Positions a view within its `Grid` cell
/// (overriding the grid's default alignment for that one cell).
export function gridCellAnchor(
  params: GridCellAnchorParams
): ReturnType<typeof createModifier>;
export function gridCellAnchor(
  anchor: GridCellAnchor
): ReturnType<typeof createModifier>;
export function gridCellAnchor(a: GridCellAnchorParams | GridCellAnchor) {
  if (typeof a === 'object' && a !== null && 'anchor' in a) {
    return createModifier('gridCellAnchor', a);
  }
  return createModifier('gridCellAnchor', { anchor: a });
}
